File: supabase/functions/_shared/ocr.ts
// ================================================================
// PureMark Edge Functions - OCR via OpenAI GPT Vision
// Uses GPT-4o/GPT-5 Vision API for text extraction from images
// ================================================================

// ================================================================
// Types
// ================================================================

export interface OCRResult {
  text: string;
  source: string;
}

// ================================================================
// GPT Vision OCR
// ================================================================

/**
 * Extract text from image using OpenAI GPT Vision API.
 * This is the primary OCR method for Supabase Edge Functions.
 *
 * @param imageData - Base64 encoded image string (with or without data URL prefix)
 * @param openaiApiKey - OpenAI API key
 * @returns Extracted text from the image
 */
export async function extractTextWithGptVision(
  imageData: string,
  openaiApiKey: string
): Promise<string> {
  // Ensure proper base64 format for OpenAI Vision API
  let imageUrl = imageData;
  if (!imageData.startsWith("data:")) {
    // Detect format from magic bytes
    let mediaType = "image/jpeg";
    if (imageData.startsWith("iVBOR")) {
      mediaType = "image/png";
    } else if (imageData.startsWith("UklGR")) {
      mediaType = "image/webp";
    }
    imageUrl = `data:${mediaType};base64,${imageData}`;
  }

  const response = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${openaiApiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: "gpt-4o",  // Use gpt-4o for best vision capabilities
      messages: [
        {
          role: "system",
          content: (
            "You are a helpful assistant that reads text from images of food product labels. " +
            "Your task is to transcribe the text exactly as it appears. This is used to help " +
            "people with dietary restrictions (halal, kosher, allergies) understand what's in their food."
          )
        },
        {
          role: "user",
          content: [
            {
              type: "text",
              text: (
                "Please read and transcribe all the text from this food ingredient label. " +
                "Include:\n" +
                "- The ingredient list\n" +
                "- Any allergen warnings (May contain, Peut contenir, etc.)\n" +
                "- Nutritional percentages if visible\n\n" +
                "Keep the original language, don't translate. Just return the text you see."
              )
            },
            {
              type: "image_url",
              image_url: {
                url: imageUrl,
                detail: "high"
              }
            }
          ]
        }
      ],
      max_tokens: 2000
    })
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`OpenAI API error: ${response.status} - ${errorText}`);
  }

  const data = await response.json();
  const extractedText = data.choices?.[0]?.message?.content?.trim() || "";

  return extractedText;
}

// ================================================================
// GPT Ingredient Parser
// ================================================================

/**
 * Parse pre-segmented ingredient zone text into structured data using GPT.
 *
 * @param ingredientZone - Pre-segmented ingredient text
 * @param detectedLanguage - Optional language code
 * @param openaiApiKey - OpenAI API key
 * @returns Parsed ingredient data
 */
export async function parseIngredientText(
  ingredientZone: string,
  detectedLanguage: string | null,
  openaiApiKey: string
): Promise<{
  detected_language: string;
  ingredients: Array<{
    original: string;
    english: string;
    normalized: string;
  }>;
  allergens: string[];
  parse_status: string;
}> {
  if (!ingredientZone || ingredientZone.trim().length < 5) {
    return {
      detected_language: detectedLanguage || "unknown",
      ingredients: [],
      allergens: [],
      parse_status: "EMPTY_ZONE"
    };
  }

  const langHint = detectedLanguage ? `The text is likely in ${detectedLanguage}. ` : "";

  const response = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${openaiApiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: "gpt-4o-mini",
      messages: [
        {
          role: "system",
          content: "Return ONLY valid JSON. You are an expert at parsing ingredient lists from OCR text."
        },
        {
          role: "user",
          content: (
            "You are given a PRE-EXTRACTED ingredient list from a food product label.\n" +
            "This text has ALREADY been segmented - it contains ONLY ingredients, " +
            "NOT product names, NOT allergen warnings, NOT nutritional info.\n\n" +
            langHint +
            "Tasks:\n" +
            "1. Confirm/detect the language\n" +
            "2. Split individual ingredients (separated by commas, semicolons, or periods)\n" +
            "3. Translate each ingredient to English ACCURATELY\n" +
            "4. Normalize ingredient names (lowercase, standard spelling)\n" +
            "5. Identify allergens PRESENT IN THE INGREDIENTS (not advisory warnings)\n\n" +
            "OCR NOISE HANDLING:\n" +
            "- Fix common OCR errors: '0' vs 'O', '1' vs 'l' vs 'I', 'rn' vs 'm'\n" +
            "- Ignore stray characters: |, *, #, @, random punctuation\n" +
            "- Ignore partial words at start/end that are clearly cut off\n" +
            "- Handle merged words (e.g., 'azucarsal' -> 'azucar, sal')\n\n" +
            "CRITICAL TRANSLATION RULES:\n" +
            "- 'tournesol' = SUNFLOWER (NOT soy)\n" +
            "- 'lécithine de tournesol' = sunflower lecithin (NOT soy lecithin)\n" +
            "- 'soja' = soy\n" +
            "- 'colza' = rapeseed/canola\n" +
            "- 'huile de palme' = palm oil\n" +
            "- 'manteca de cacao' / 'beurre de cacao' = cocoa butter\n" +
            "- 'pasta de cacao' / 'pâte de cacao' = cocoa mass\n" +
            "- 'azúcar' / 'sucre' = sugar\n" +
            "- 'lait' / 'leche' = milk\n" +
            "- 'oeuf' / 'oeufs' / 'huevo' = egg\n" +
            "- 'blé' / 'trigo' = wheat\n" +
            "- 'arachide' / 'cacahuete' = peanut\n" +
            "- Do NOT assume lecithin is soy unless the original explicitly says 'soja' or 'soy'\n" +
            "- Keep lecithin source accurate in translation\n\n" +
            "WHAT IS NOT AN INGREDIENT (ignore if present):\n" +
            "- Percentages like 'Cacao: 78%' or '85% cacao'\n" +
            "- Weight/volume like '100g', '200ml'\n" +
            "- Product names like 'Chocolate Negro'\n" +
            "- Certifications like 'UTZ', 'Fairtrade'\n" +
            "- 'May contain' warnings\n\n" +
            "Return JSON:\n" +
            "{\n" +
            "  \"detected_language\": \"<language code: en, fr, de, es, it, etc.>\",\n" +
            "  \"ingredients\": [\n" +
            "    { \"original\": \"<original text>\", \"english\": \"<english translation>\", \"normalized\": \"<normalized lowercase>\" }\n" +
            "  ],\n" +
            "  \"allergens\": [\"<allergen1>\", \"<allergen2>\"]\n" +
            "}\n\n" +
            `Ingredient zone text:\n${ingredientZone}`
          )
        }
      ],
      max_tokens: 600
    })
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`OpenAI API error: ${response.status} - ${errorText}`);
  }

  const data = await response.json();
  const raw = data.choices?.[0]?.message?.content || "";

  // Parse JSON from response
  const result = safeExtractJson(raw);
  result.parse_status = "OK";
  return result;
}

/**
 * Safely extract JSON from model output that may contain markdown formatting.
 */
function safeExtractJson(raw: string): any {
  // Remove markdown code blocks
  let cleaned = raw.replace(/```(?:json)?/g, "").replace(/```/g, "").trim();

  // Find JSON object
  const match = cleaned.match(/\{[\s\S]*\}/);
  if (!match) {
    throw new Error("Model did not return a JSON object.");
  }

  return JSON.parse(match[0]);
}

File: backend/services/ocr.py
# ================================================================
# PureMark Backend - OCR Service
# Gemini 2.0 Flash via OpenRouter for text extraction
# GPT-4o-mini for parsing
# ================================================================

import httpx
import json
import base64
from typing import Dict, Any


async def extract_text_with_gemini(image_base64: str, api_key: str) -> str:
    """
    Extract text from image using Gemini 2.0 Flash via OpenRouter.

    Args:
        image_base64: Base64 encoded image
        api_key: OpenRouter API key

    Returns:
        Extracted text from the image
    """

    # Detect image type from base64 header
    media_type = "image/jpeg"  # default
    if image_base64.startswith("/9j/"):
        media_type = "image/jpeg"
    elif image_base64.startswith("iVBOR"):
        media_type = "image/png"
    elif image_base64.startswith("UklGR"):
        media_type = "image/webp"

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
        "HTTP-Referer": "https://puremark.app",
        "X-Title": "PureMark"
    }

    payload = {
        "model": "google/gemini-2.0-flash-001",
        "max_tokens": 2000,
        "messages": [
            {
                "role": "user",
                "content": [
                    {
                        "type": "image_url",
                        "image_url": {
                            "url": f"data:{media_type};base64,{image_base64}"
                        }
                    },
                    {
                        "type": "text",
                        "text": """You are an OCR specialist. Extract ALL text from this food label image exactly as written.
Focus on the ingredients list section. Preserve the original language and formatting.
Return ONLY the raw text, no commentary or formatting."""
                    }
                ]
            }
        ]
    }

    async with httpx.AsyncClient(timeout=60.0) as client:
        response = await client.post(
            "https://openrouter.ai/api/v1/chat/completions",
            headers=headers,
            json=payload
        )
        response.raise_for_status()

        data = response.json()
        return data["choices"][0]["message"]["content"]


# Keep the old function name as alias for compatibility
async def extract_text_with_gpt_vision(image_base64: str, api_key: str) -> str:
    """Alias - now uses Gemini 2.0 Flash via OpenRouter."""
    return await extract_text_with_gemini(image_base64, api_key)


async def parse_ingredient_text(
    ingredient_text: str,
    detected_language: str,
    api_key: str
) -> Dict[str, Any]:
    """
    Parse and structure ingredient text using GPT-4o-mini.

    Args:
        ingredient_text: Raw ingredient text
        detected_language: Detected language of text
        api_key: OpenAI API key

    Returns:
        Parsed ingredients with translations
    """

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }

    payload = {
        "model": "gpt-4o-mini",
        "messages": [
            {
                "role": "system",
                "content": """You are a food ingredient parser. Parse the ingredient list and return JSON.

For each ingredient:
1. "original": The ingredient as written (preserve original language)
2. "english": English translation (if not already English)
3. "normalized": Lowercase, cleaned English name for matching

Also extract:
- "detected_language": The language of the ingredients
- "allergens": Any allergens mentioned (peanuts, tree nuts, milk, eggs, fish, shellfish, soy, wheat, sesame)

Return valid JSON only, no markdown."""
            },
            {
                "role": "user",
                "content": f"""Parse these ingredients (detected language: {detected_language}):

{ingredient_text}

Return JSON with format:
{{
  "detected_language": "string",
  "ingredients": [
    {{"original": "string", "english": "string", "normalized": "string"}}
  ],
  "allergens": ["string"]
}}"""
            }
        ],
        "max_tokens": 2000,
        "response_format": {"type": "json_object"}
    }

    async with httpx.AsyncClient(timeout=60.0) as client:
        response = await client.post(
            "https://api.openai.com/v1/chat/completions",
            headers=headers,
            json=payload
        )
        response.raise_for_status()

        data = response.json()
        content = data["choices"][0]["message"]["content"]

        # Parse JSON response
        try:
            parsed = json.loads(content)
            return parsed
        except json.JSONDecodeError:
            # Try to extract JSON from response
            import re
            json_match = re.search(r'\{.*\}', content, re.DOTALL)
            if json_match:
                return json.loads(json_match.group())
            raise ValueError("Could not parse ingredient response")
